expan/core/results.py

```diff
@@ -0,0 +1,110 @@
+    :type statistical_power: float
+    :type ci: ConfidenceInterval
+    """
+    def __init__(self, control_statistics, treatment_statistics, delta, ci, p, statistical_power):
+        super(SimpleTestStatistics, self).__init__(control_statistics, treatment_statistics)
+        self.delta               = delta
+        self.p                   = p
+        self.statistical_power   = statistical_power
+        self.confidence_interval = ci
+
+
+class EarlyStoppingTestStatistics(SimpleTestStatistics):
+    """ Additionally to SimpleTestStatistics, holds boolean flag for early stopping.
+    :type control_statistics: SampleStatistics
+    :type treatment_statistics: SampleStatistics
+    :type delta: float
+    :type p: float
+    :type statistical_power: float
+    :type ci: ConfidenceInterval
+    :type stop: bool
+    """
+    def __init__(self, control_statistics, treatment_statistics, delta, ci, p, statistical_power, stop):
+        super(EarlyStoppingTestStatistics, self).__init__(control_statistics, treatment_statistics, delta, ci, p, statistical_power)
+        self.stop = stop
+
+
+class CorrectedTestStatistics(object):
+    """ Holds original and corrected statistics. This class should be used to hold statistics for multiple testing.
+    original_test_statistics and corrected_test_statistics should have the same type.
+    :param original_test_statistics: test result before correction
+    :type  original_test_statistics: SimpleTestStatistics or EarlyStoppingTestStatistics
+    :param corrected_test_statistics: test result after correction
+    :type  corrected_test_statistics: SimpleTestStatistics or EarlyStoppingTestStatistics
+    """
+    def __init__(self, original_test_statistics, corrected_test_statistics):
+        type1 = type(original_test_statistics)
+        type2 = type(corrected_test_statistics)
+        if type1 != type2:
+            raise RuntimeError("Type mismatch for type " + str(type1) + " and " + str(type2))
+        self.original_test_statistics  = original_test_statistics
+        self.corrected_test_statistics = corrected_test_statistics
+
+
+class ConfidenceInterval(object):
+    """ This class represents the confidence interval. 
+    :param confidence_interval: a dict where keys are percentiles and values are the corresponding value for the statistic.
+    :type  confidence_interval: dict
+    """
+    def __init__(self, confidence_interval):
+        c_i = [{'percentile': p, 'value': v} for (p, v) in confidence_interval.items()]
+        self.confidence_interval = c_i
+
+
+# --------- Below are the data structure of test results --------- #
+class StatisticalTestResult(object):
+    """ This class holds the results of a single statistical test.
+    :param test: information about the statistical test
+    :type  test: StatisticalTest
+    :param result: result of this statistical test
+    :type  result: BaseTestStatistics or its subclasses or CorrectedTestStatistics  #TODO: better approach?
```

Comment:
Here is another place I found our design can be improved.

Reply:
@shansfolder yes, I feel it too, but don't have an immediate solution now.

I think it's fine for the time being, since we can change the implementation details later and leave the interface untouched by introducing/redefining `test` and `results` properties.